core/src/main/java/zipkin/reporter/AsyncReporter.java

```diff
@@ -0,0 +1,253 @@
+     * ReporterMetrics#forTransport(String) scoped to this transport}.  Defaults to no-op.
+     */
+    public Builder<S> metrics(ReporterMetrics metrics) {
+      this.metrics = checkNotNull(metrics, "metrics");
+      return this;
+    }
+
+    /** Maximum backlog of spans reported vs sent. Default 1000 */
+    public Builder maxQueueDepth(int maxQueueDepth) {
+      this.maxQueueDepth = maxQueueDepth;
+      return this;
+    }
+
+    /** Maximum count of bytes allowed in a message batch. */
+    public Builder maxBatchByteCount(int maxBatchByteCount) {
+      this.maxBatchByteCount = maxBatchByteCount;
+      return this;
+    }
+
+    /** Builds an async reporter that does not encode spans as they are reported. */
+    public AsyncReporter<S> build() {
+      checkArgument(maxBatchByteCount == null, "maxBatchByteCount requires encoder");
+      BoundedQueue<S> queue = BoundedQueue.create(maxQueueDepth, maxBatchSize);
+      return BoundedAsyncReporter.create(this, queue);
+    }
+
+    /** Builds an async reporter that encodes spans as they are reported. */
+    public <S1> AsyncReporter<S1> build(Encoder<S1, S> encoder) {
+      checkNotNull(encoder, "encoder");
+      checkArgument(sender.encoding() == Encoding.OPAQUE || encoder.encoding() == sender.encoding(),
+          "Sender.encoding() %s != Encoder.encoding() %s", sender.encoding(), encoder.encoding());
+      BoundedQueue<S> queue;
+      if (maxBatchByteCount != null) {
+        queue = BoundedQueue.create(maxQueueDepth, encoder::byteCount, maxBatchByteCount);
+      } else {
+        queue = BoundedQueue.create(maxQueueDepth, maxBatchSize);
+      }
+      BoundedAsyncReporter<S> delegate = BoundedAsyncReporter.create(this, queue);
+      return new AutoValue_AsyncReporter_EncodingAsyncReporter<>(
+          encoder,
+          delegate,
+          metrics,
+          delegate.closed()
+      );
+    }
+  }
+
+  @AutoValue
+  static abstract class BoundedAsyncReporter<S> extends AsyncReporter<S> {
+    static final Logger logger = Logger.getLogger(BoundedAsyncReporter.class.getName());
+
+    abstract BoundedQueue<S> pending();
+
+    abstract Sender<S> sender();
+
+    abstract ReporterMetrics metrics();
+
+    abstract AtomicBoolean closed();
+
+    abstract CountDownLatch shutdown();
+
+    static <S> BoundedAsyncReporter<S> create(Builder<S> builder, BoundedQueue<S> queue) {
+      AtomicBoolean closed = new AtomicBoolean(false);
+      CountDownLatch shutdown = new CountDownLatch(1);
+      BoundedAsyncReporter<S> result = new AutoValue_AsyncReporter_BoundedAsyncReporter<>(
+          queue,
+          builder.sender,
+          builder.metrics,
+          closed,
+          shutdown
+      );
+      new Thread(() -> {
+        try {
+          while (!closed.get()) {
+            result.flush();
+          }
+        } finally {
+          shutdown.countDown();
+        }
+      }, "Flusher(" + builder.sender + ")").start();
```

Comment:
I know daemon threads are a polarizing topic.. if we make this daemon, then people could forget to call close and process shutdown will silently toss things.. If we don't make this daemon, someone who forgets to call close will have their process hung and even looking at this thread name.. maybe they will still complain.

Any thoughts?